### PR TITLE
Feature/implement generic add

### LINKED ISSUE #1427: Feat: multi add function to add multiple parts of a date
Similar to #1388 (which was implemented by @SneakyFish5 in #1398 ), it would be nice to have a function, that can add multiple values at once.

So instead of doing this:
```
addSeconds(addMinutes(addHours(myDate, 12), 12), 12);
```

Something like this would be nice:
```
add(myDate, { hours: 12, minutes: 12, seconds: 12 });
```

### PR DESCRIPTION
**Pull request by Anshuman71**
Implement the `add(Object)` function from [Moment](https://momentjs.com/docs/#/manipulating/add/) and fix #1427 

> Add 2 hours, 15 minutes and 30 seconds to 10 July 2014 12:45:00;
```
var result = add(new Date(2014, 6, 10, 12, 45, 0), { minutes: 15, hours: 2, seconds: 30 });
```
> result: Thu Jul 10 2014 15:00:30

Supports adding 
```
{
    hours: number,
    years: number,
    days: number,
    quarters: number,
    weeks: number,
    months: number,
    minutes: number,
    seconds: number
} 
```
combined in a single call

### CODE REVIEW

**Comment by Anshuman71** [comment]
@leshakoss @alaskaa @dkozickis can you please review this PR

File: src/add/index.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      '2 arguments required, but only ' + arguments.length + ' present'
+    )
+  }
+  if (typeof givenAmount !== 'object' || !Object.keys(givenAmount).length) {
+    throw new TypeError(
+      '2nd argument expected Object, found ' + typeof givenAmount
+    )
+  }
+  const givenDate = toDate(dirtyDate)
+  const dirtyAmount = Object.assign({}, DEFAULT)
```
Comment (dkozickis) [inline src/add/index.js]: Using manual check instead of `Object.assign` and destructing will do wonders for performance here - https://jsperf.com/assign-vs-manual-check/1

Consider changing to something like 

```javascript
const seconds = givenAmount.seconds || 0;
const minutes = givenAmount.minutes || 0;
const hours = givenAmount.hours || 0;
const days = givenAmount.days || 0;
const months = givenAmount.months || 0;
const quarters = givenAmount.quarters || 0;
const weeks = givenAmount.weeks || 0;
const years = givenAmount.years || 0;
```

File: src/add/index.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    hours,
+    years,
+    days,
+    quarters,
+    weeks,
+    months,
+    minutes,
+    seconds
+  } = dirtyAmount
+  return addYears(
```
Comment (dkozickis) [inline src/add/index.js]: This can be optimized by skipping time span values that are null and changing `givenDate`

```javascript
let givenDate = toDate(dirtyDate);

givenDate = years ? addYears(givenDate, toInteger(years)) : givenData;
givenDate = months ? addMonths(givenDate, toInteger(months)) : givenDate;
```

File: src/add/test.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+// @flow
+/* eslint-env mocha */
+
+import assert from 'power-assert'
+import add from '.'
+
+describe('add', function() {
```
Comment (dkozickis) [inline src/add/test.js]: Please add test with empty options object and with full options object (including all values like `months`, `years`...)

**Comment by Anshuman71** [comment]
> The function is good, but as of now in a bundle it would be huge: it includes, for example, the impossible condition `if (arguments.length < 2) ...` 8 times.
> 
> Please optimize the function by copying and pasting the logic inside all of `addX` functions into this one, optimizing where possible (for example you can add ms, seconds, minutes and hours on the same line because all of these are milliseconds multiplied by some constant). After that, the function should be able to do everything `addX` do, but without importing them.

Thankyou, I will surely optimize the code with the above advise

File: src/add/test.js (hunk 3 above)
Comment (Anshuman71) [reply src/add/test.js]: Sure, sounds good

File: src/add/index.js (hunk 2 above)
Comment (Anshuman71) [reply src/add/index.js]: Yeah, this would help increase efficiency

File: src/add/index.js (hunk 1 above)
Comment (Anshuman71) [reply src/add/index.js]: Thankyou for pointing it out. I learned something new.

File: src/add/index.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * @name add
+ * @category Combined time Helpers
+ * @summary Add the specified seconds, minutes, hours, days, months, quarters and years to the given date.
+ *
+ * @description
+ * Add the specified seconds, minutes, hours, days, months, quarters and years to the given date.
+ *
+ * ### v2.0.0 breaking changes:
```
[the remark this answers is not in the document]
Comment (Anshuman71) [reply src/add/index.js]: Sorry for that, I am new in here. Will take care next time. Thanks for pointing out

File: CHANGELOG.md [hunk 5]
```diff
@@ -13,7 +13,7 @@ This change log follows the format documented in [Keep a CHANGELOG].
 Kudos to [@marnusw](https://github.com/marnusw), [@cdrikd](https://github.com/cdrikd) and [@rogyvoje](https://github.com/rogyvoje) for working on the release!
 
 ### Added
-
+- [Added `add` - function to add seconds, minutes, hours, weeks, years in single call](https://github.com/date-fns/date-fns/pull/1504)
```
[the remark this answers is not in the document]
Comment (Anshuman71) [reply CHANGELOG.md]: Okay, I will do

File: src/add/index.js (hunk 1 above)
Comment (Anshuman71) [reply src/add/index.js]: @dkozickis  But in this case we will have to check for every key if they are NaN to return `Invalid Date` otherwise it will resolve to 0 and return a valid date

**Comment by kossnocorp** [comment]
@Anshuman71 hey, I love the PR, but I see that I've removed the repo, and now I can't resolve the conflicts and clean up the code. I'll appreciate it if you can push it again or send me the patch file. Thanks!

**Comment by Anshuman71** [comment]
> @Anshuman71 hey, I love the PR, but I see that I've removed the repo, and now I can't resolve the conflicts and clean up the code. I'll appreciate it if you can push it again or send me the patch file. Thanks!

**Comment by Anshuman71** [comment]
Hey @kossnocorp, thanks for your feedback. I moved this to #1581 . Please look into and let me know if need any changes.
